File: Day35/Part1/min_subarray_removal_divisible_p.py
from typing import List

def min_subarray(nums: List[int], p: int) -> int:
    total = sum(nums)
    target = total % p
    if target == 0:
        return 0
    seen = {0: -1}
    res = len(nums)
    cur = 0
    for i, x in enumerate(nums):
        cur = (cur + x) % p
        need = (cur - target) % p
        if need in seen:
            res = min(res, i - seen[need])
        seen[cur] = i
    return res if res < len(nums) else -1

if __name__ == "__main__":
    print(min_subarray([3,1,4,2], 6))  # Expected 1 (remove [4])

File: learnLeetCode_remote/Day35/Part1/min_subarray_removal_divisible_p.cpp
#include <bits/stdc++.h>
using namespace std;

int min_subarray(vector<int> nums, int p){
    long long total = 0; for(int x: nums) total += x;
    int target = (int)(total % p);
    if(target == 0) return 0;
    unordered_map<int,int> pos; pos.reserve(nums.size()*2); pos[0] = -1;
    int res = (int)nums.size();
    long long cur = 0;
    for(int i=0;i<(int)nums.size();++i){
        cur = (cur + nums[i]) % p;
        int need = (int)((cur - target) % p);
        if(need < 0) need += p;
        auto it = pos.find(need);
        if(it != pos.end()) res = min(res, i - it->second);
        pos[(int)cur] = i;
    }
    return res < (int)nums.size() ? res : -1;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << min_subarray(vector<int>{3,1,4,2}, 6) << "\n"; // Expected 1
    return 0;
}
